src/hotspot/share/gc/z/zObjArrayAllocator.cpp

```diff
@@ -88,6 +75,22 @@ oop ZObjArrayAllocator::initialize(HeapWord* mem) const {
   // over such objects.
   ZThreadLocalData::set_invisible_root(_thread, (zaddress_unsafe*)&mem);
 
+  const BasicType element_type = ArrayKlass::cast(_klass)->element_type();
+  const size_t base_offset_in_bytes = arrayOopDesc::base_offset_in_bytes(element_type);
+  const size_t process_start_offset_in_bytes = align_up(base_offset_in_bytes, BytesPerWord);
+
+  if (process_start_offset_in_bytes != base_offset_in_bytes) {
+    // initialize_memory can only fill word aligned memory,
+    // fill the first 4 bytes here.
+    assert(process_start_offset_in_bytes - base_offset_in_bytes == 4, "Must be 4-byte aligned");
+    assert(!is_reference_type(element_type), "Only TypeArrays can be 4-byte aligned");
```

Comment:
Actually, only the type arrays whose element type is not double word can be 4-byte aligned. So this assertion could be `!is_reference_type(element_type) && !is_double_word_type(element_type)`.

Reply:
I think the point is that when it is not a reference type, it is OK to initialize with a plain `0`.

Reply:
May be better to treat the assertion as a more detailed guard or check. But I'm also OK with the current code.